File: ods/extensions/services/pixel-agent/plugin/workspace-bundle.py
"""Fixed workspace-only byte bundler; invoked by the trusted plugin adapter.

The model supplies paths and keys, never code or file contents. Each
execution creates a unique generation and never replaces an existing output.
A failed generation is retained as incomplete, never reported as verified.
"""
import base64
import hashlib
import json
import os
import re
import secrets
import stat
import sys

MAX_FILE = 256 * 1024
MAX_TOTAL = 1024 * 1024
FLAGS = os.O_RDONLY | getattr(os, 'O_NOFOLLOW', 0) | getattr(os, 'O_NONBLOCK', 0)
COMPONENT = re.compile(r"[A-Za-z0-9][A-Za-z0-9._ -]{0,127}")
SHA = re.compile(r"[a-f0-9]{64}")


class BundleError(Exception):
    pass


def require(condition, reason):
    if not condition:
        raise BundleError(reason)


def digest(data):
    return hashlib.sha256(data).hexdigest()


def parts(value):
    require(isinstance(value, str) and 0 < len(value) <= 512, "invalid-path")
    result = value.split('/')
    require(len(result) <= 16 and all(COMPONENT.fullmatch(x) and not x.endswith(('.', ' '))
            and x not in ('.', '..') for x in result), "invalid-path")
    return result


def validate(request):
    require(isinstance(request, dict) and set(request) == {'files', 'mappingPath', 'outputRoot'}, "invalid-request")
    files = request['files']
    require(isinstance(files, list) and 1 <= len(files) <= 32, "invalid-file-count")
    parts(request['mappingPath'])
    parts(request['outputRoot'])
    sources, outputs, keys = set(), {request['mappingPath'].casefold()}, set()
    for item in files:
        require(isinstance(item, dict) and set(item) == {'source', 'key', 'copyTo'}, "invalid-file")
        parts(item['source'])
        parts(item['copyTo'])
        require(isinstance(item['key'], str) and 0 < len(item['key']) <= 256
                and not any(ord(c) < 32 for c in item['key']), "invalid-key")
        require(item['key'] not in keys and item['source'].casefold() not in sources
                and item['copyTo'].casefold() not in outputs, "duplicate-path-or-key")
        keys.add(item['key'])
        sources.add(item['source'].casefold())
        outputs.add(item['copyTo'].casefold())
    all_paths = sorted(outputs)
    require(not any(b.startswith(a + '/') for a, b in zip(all_paths, all_paths[1:])), "nested-file-path")
    return files


def stable(info):
    return (info.st_dev, info.st_ino, info.st_size, info.st_mtime_ns, info.st_ctime_ns)


def parent(root, relative, create=False):
    fd = os.dup(root)
    try:
        for component in parts(relative)[:-1]:
            if create:
                try:
                    os.mkdir(component, 0o700, dir_fd=fd)
                except FileExistsError:
                    pass
            child = os.open(component, FLAGS | os.O_DIRECTORY, dir_fd=fd)
            os.close(fd)
            fd = child
        return fd
    except BaseException:
        os.close(fd)
        raise


def read(root, relative, limit=MAX_FILE, missing=False, with_identity=False):
    try:
        directory = parent(root, relative)
    except FileNotFoundError:
        if missing:
            return None
        raise
    try:
        try:
            fd = os.open(parts(relative)[-1], FLAGS, dir_fd=directory)
        except FileNotFoundError:
            if missing:
                return None
            raise
        try:
            before = os.fstat(fd)
            require(stat.S_ISREG(before.st_mode) and before.st_nlink == 1 and before.st_size <= limit, "unsafe-file")
            data = bytearray()
            while len(data) <= limit:
                block = os.read(fd, min(65536, limit + 1 - len(data)))
                if not block:
                    break
                data.extend(block)
            require(len(data) <= limit and stable(before) == stable(os.fstat(fd)), "changed-or-large-file")
            require(stable(before) == stable(os.stat(parts(relative)[-1], dir_fd=directory, follow_symlinks=False)), 'source-path-changed')
            return (bytes(data), stable(before)) if with_identity else bytes(data)
        finally:
            os.close(fd)
    finally:
        os.close(directory)


def derive(root, request):
    files = validate(request)
    sources, outputs, mapping, identities, total = [], {}, {}, [], 0
    for item in files:
        data, identity = read(root, item['source'], with_identity=True)
        identities.append(identity)
        total += len(data)
        require(total <= MAX_TOTAL, "source-total-too-large")
        try:
            text = data.decode('utf-8', errors='strict')
        except UnicodeDecodeError:
            raise BundleError('source-not-utf8')
        require(text.encode('utf-8') == data, "source-roundtrip-failed")
        sources.append({'path': item['source'], 'bytes': len(data), 'sha256': digest(data)})
        mapping[item['key']] = text
        outputs[item['copyTo']] = data
    # JSON whitespace is deterministic; the values preserve every original byte.
    encoded = (json.dumps(mapping, ensure_ascii=False, sort_keys=True, separators=(',', ':')) + '\n').encode('utf-8')
    parsed = json.loads(encoded)
    require(all(parsed[item['key']].encode('utf-8') == outputs[item['copyTo']] for item in files), "mapping-roundtrip-failed")
    outputs[request['mappingPath']] = encoded
    return sources, outputs, identities


def atomic_write(root, name, data):
    directory = parent(root, name, create=True)
    temporary = '.ods-bundle-' + secrets.token_hex(12)
    fd = None
    try:
        fd = os.open(temporary, os.O_RDWR | os.O_CREAT | os.O_EXCL | os.O_NOFOLLOW, 0o600, dir_fd=directory)
        offset = 0
        while offset < len(data):
            offset += os.write(fd, data[offset:])
        os.fsync(fd)
        os.lseek(fd, 0, os.SEEK_SET)
        observed = bytearray()
        while len(observed) < len(data):
            chunk = os.read(fd, min(65536, len(data) - len(observed)))
            require(bool(chunk), "staged-readback-failed")
            observed.extend(chunk)
        require(bytes(observed) == data, "staged-readback-failed")
        os.close(fd)
        fd = None
        rebound = parent(root, name)
        try:
            current_parent, original_parent = os.fstat(rebound), os.fstat(directory)
            require((current_parent.st_dev, current_parent.st_ino) ==
                    (original_parent.st_dev, original_parent.st_ino), "output-parent-changed")
        finally:
            os.close(rebound)
        # Atomic create-only publication: a concurrent owner/editor-created
        # destination wins, even when it appears after all prior checks.
        os.link(temporary, parts(name)[-1], src_dir_fd=directory, dst_dir_fd=directory, follow_symlinks=False)
        os.unlink(temporary, dir_fd=directory)
        os.fsync(directory)
    finally:
        if fd is not None:
            os.close(fd)
        try:
            os.unlink(temporary, dir_fd=directory)
        except FileNotFoundError:
            pass
        os.close(directory)


def bundle(payload, workspace='.'):
    require(os.name == 'posix' and hasattr(os, 'O_NOFOLLOW'), 'posix-workspace-required')
    require(isinstance(payload, dict) and set(payload) == {'request', 'generation'}, "invalid-operation")
    request, generation = payload['request'], payload['generation']
    validate(request)
    require(isinstance(generation, str) and re.fullmatch(r'bundle-[a-f0-9]{32}', generation), 'invalid-generation')
    generation_path = request['outputRoot'] + '/' + generation
    parts(generation_path)
    root = os.open(workspace, FLAGS | os.O_DIRECTORY)
    written, created, directory = [], False, None
    try:
        sources, outputs, identities = derive(root, request)
        holder = parent(root, generation_path, create=True)
        try:
            os.mkdir(generation, 0o700, dir_fd=holder)
            created = True
            directory = os.open(generation, FLAGS | os.O_DIRECTORY, dir_fd=holder)
        finally:
            os.close(holder)
        for name, data in outputs.items():
            atomic_write(directory, name, data)
            written.append(generation_path + '/' + name)
        require(all(read(directory, name, MAX_TOTAL * 6 + 4096) == data for name, data in outputs.items()), 'output-readback-failed')
        final_sources, _, final_identities = derive(root, request)
        require(final_sources == sources and final_identities == identities, 'source-changed-during-generation')
        # Reopen through the workspace path too; an externally renamed parent
        # must never turn an unreachable generation into a success receipt.
        require(all(read(root, generation_path + '/' + name, MAX_TOTAL * 6 + 4096) == data for name, data in outputs.items()), 'generation-path-changed')
        return {'schemaVersion': 2, 'kind': 'ods-pixel-workspace-bundle', 'status': 'succeeded',
                'generationPath': generation_path, 'mappingPath': generation_path + '/' + request['mappingPath'],
                'sources': sources, 'outputs': [{'path': generation_path + '/' + name, 'bytes': len(data), 'sha256': digest(data)} for name, data in sorted(outputs.items())],
                'written': written, 'readbackVerified': True, 'decodedAndRawBytesEqual': True,
                'boundary': 'Immutable derived bytes only; no execution, publication, or task-correctness proof.'}
    except BaseException as error:
        return {'schemaVersion': 2, 'kind': 'ods-pixel-workspace-bundle', 'status': 'failed',
                'generationPath': generation_path if created else None,
                'written': written, 'readbackVerified': False,
                'error': str(error) if isinstance(error, BundleError) else 'workspace-io-failed'}
    finally:
        if directory is not None:
            os.close(directory)
        os.close(root)


if __name__ == '__main__':
    try:
        require(len(sys.argv) == 2 and len(sys.argv[1]) <= 131072, 'invalid-payload')
        value = bundle(json.loads(base64.b64decode(sys.argv[1], validate=True)))
    except BaseException:
        value = {'schemaVersion': 2, 'kind': 'ods-pixel-workspace-bundle', 'status': 'failed', 'error': 'invalid-payload', 'readbackVerified': False}
    print(json.dumps(value, ensure_ascii=True, separators=(',', ':')))
    raise SystemExit(0 if value['status'] == 'succeeded' else 1)

File: ods/extensions/services/pixel-agent/plugin/workspace-bundle.mjs
import * as fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import {createHash, randomUUID} from 'node:crypto';
import {isDeepStrictEqual} from 'node:util';

export const WORKSPACE_BUNDLE_TOOL = 'pixel_ods_workspace_bundle';
const hash = value => createHash('sha256').update(value).digest('hex');
const sha = value => typeof value === 'string' && /^[a-f0-9]{64}$/.test(value);
const fail = code => Object.assign(new Error(code), {code});
const bundleQueues = new Map();
const exact = (value, keys) => value && typeof value === 'object' && !Array.isArray(value) &&
  Object.keys(value).sort().join() === keys.split(',').sort().join();
const safePath = value => typeof value === 'string' && value.length <= 512 && value.split('/').length <= 16 &&
  value.split('/').every(part => /^[A-Za-z0-9][A-Za-z0-9._ -]{0,127}$/.test(part) && !/[. ]$/.test(part));

export function normalizeWorkspaceBundle(value) {
  if (!exact(value, 'files,mappingPath,outputRoot') || !safePath(value.outputRoot) || !safePath(value.mappingPath) ||
      !Array.isArray(value.files) || value.files.length < 1 || value.files.length > 32) throw fail('invalid-bundle-request');
  const sources = new Set(), outputs = new Set([value.mappingPath.toLowerCase()]), keys = new Set();
  for (const item of value.files) {
    if (!exact(item, 'source,key,copyTo') || !safePath(item.source) || !safePath(item.copyTo) ||
        typeof item.key !== 'string' || !item.key || item.key.length > 256 || /[\x00-\x1f]/.test(item.key) ||
        sources.has(item.source.toLowerCase()) || outputs.has(item.copyTo.toLowerCase()) || keys.has(item.key)) throw fail('invalid-bundle-file');
    sources.add(item.source.toLowerCase()); outputs.add(item.copyTo.toLowerCase()); keys.add(item.key);
  }
  const paths = [...outputs].sort();
  if (paths.some((name, index) => paths[index + 1]?.startsWith(name + '/'))) throw fail('nested-file-path');
  if (paths.some(name => !safePath(`${value.outputRoot}/bundle-${'a'.repeat(32)}/${name}`))) throw fail('bundle-path-too-long');
  return JSON.parse(JSON.stringify(value));
}

function helperReceipt(value, request, generation) {
  const prefix = `${request.outputRoot}/${generation}/`;
  const outputs = [request.mappingPath, ...request.files.map(item => item.copyTo)].sort().map(name => prefix + name);
  if (value?.schemaVersion !== 2 || value.kind !== 'ods-pixel-workspace-bundle' || value.status !== 'succeeded' ||
      value.generationPath !== prefix.slice(0,-1) || value.mappingPath !== prefix + request.mappingPath ||
      value.readbackVerified !== true || value.decodedAndRawBytesEqual !== true ||
      !Array.isArray(value.sources) || !Array.isArray(value.outputs) ||
      !isDeepStrictEqual(value.sources.map(item => item.path), request.files.map(item => item.source)) ||
      !isDeepStrictEqual(value.outputs.map(item => item.path), outputs) ||
      [...value.sources, ...value.outputs].some(item => !exact(item, 'path,bytes,sha256') ||
        !Number.isSafeInteger(item.bytes) || item.bytes < 0 || item.bytes > 6 * 1024 * 1024 + 4096 || !sha(item.sha256)) ||
      value.sources.some(item => item.bytes > 256 * 1024) ||
      value.sources.reduce((total, item) => total + item.bytes, 0) > 1024 * 1024 ||
      request.files.some((item, index) => {
        const source = value.sources[index], output = value.outputs.find(out => out.path === prefix + item.copyTo);
        return !output || output.bytes !== source.bytes || output.sha256 !== source.sha256;
      })) throw fail('unverified-bundle-receipt');
  return value;
}

// Records only an actually admitted tool call, including Tool Search's exact
// parent/child ID convention. The factory context alone has no live run ID.
export function createWorkspaceBundleAdmission() {
  const pending = new Map();
  return {
    before(event, context, decision) {
      if (decision?.block || context?.agentId !== 'pixel') return;
      const params = decision?.params ?? event?.params;
      const direct = event?.toolName === WORKSPACE_BUNDLE_TOOL;
      const deferred = event?.toolName === 'tool_call' &&
        [WORKSPACE_BUNDLE_TOOL, 'openclaw:pixel-ods:' + WORKSPACE_BUNDLE_TOOL].includes(params?.id);
      if (!direct && !deferred) return;
      const id = context?.toolCallId ?? event?.toolCallId;
      if (!id || !context?.runId || !context?.sessionId || !context?.sessionKey) return;
      while (pending.size >= 128) pending.delete(pending.keys().next().value);
      try { pending.set(id, {context: {...context}, request: normalizeWorkspaceBundle(deferred ? params.args : params), deferred}); }
      catch { /* The actual tool reports invalid arguments without admitting execution. */ }
    },
    take(id, request, factory) {
      let entry = pending.get(id);
      if (!entry) {
        const matches = [...pending].filter(([parentId, item]) => {
          const parent = parentId.trim().replace(/[^A-Za-z0-9_.:-]+/g, '_').slice(0, 120) || 'call';
          const prefix = `tool_search_code:${parent}:${WORKSPACE_BUNDLE_TOOL}:`;
          return item.deferred && id.startsWith(prefix) && /^[1-9][0-9]*$/.test(id.slice(prefix.length));
        });
        if (matches.length === 1) entry = matches[0][1];
      }
      if (!entry || entry.used || !isDeepStrictEqual(entry.request, request) ||
          ['agentId', 'sessionId', 'sessionKey'].some(key => entry.context[key] !== factory?.[key])) throw fail('bundle-call-unbound');
      entry.used = true;
      return Object.freeze({...entry.context});
    },
    after(event, context) { pending.delete(context?.toolCallId ?? event?.toolCallId); },
  };
}

function privateDirectory(directory) {
  if (!path.isAbsolute(directory)) throw fail('bundle-state-unavailable');
  fs.mkdirSync(directory, {recursive: true, mode: 0o700});
  const info = fs.lstatSync(directory);
  if (!info.isDirectory() || info.isSymbolicLink() || fs.realpathSync(directory) !== directory ||
      process.platform !== 'win32' && (info.uid !== process.getuid() || (info.mode & 0o077))) throw fail('bundle-state-unavailable');
}

function readPrivate(file) {
  let before;
  try { before = fs.lstatSync(file); } catch (error) { if (error.code === 'ENOENT') return null; throw error; }
  if (!before.isFile() || before.isSymbolicLink() || before.nlink !== 1 || before.size > 128 * 1024 ||
      process.platform !== 'win32' && (before.uid !== process.getuid() || (before.mode & 0o077))) throw fail('bundle-state-unavailable');
  const fd = fs.openSync(file, fs.constants.O_RDONLY | (fs.constants.O_NOFOLLOW ?? 0));
  try {
    const opened = fs.fstatSync(fd);
    if (opened.ino !== before.ino || opened.dev !== before.dev) throw fail('bundle-state-changed');
    return JSON.parse(fs.readFileSync(fd, 'utf8'));
  } finally { fs.closeSync(fd); }
}

function savePrivate(file, value) {
  const temporary = file + '.' + randomUUID();
  const fd = fs.openSync(temporary, fs.constants.O_WRONLY | fs.constants.O_CREAT | fs.constants.O_EXCL | (fs.constants.O_NOFOLLOW ?? 0), 0o600);
  try { fs.writeFileSync(fd, JSON.stringify(value)); fs.fsyncSync(fd); }
  finally { fs.closeSync(fd); }
  try { fs.renameSync(temporary, file); }
  finally { try { fs.unlinkSync(temporary); } catch (error) { if (error.code !== 'ENOENT') throw error; } }
}

export function createWorkspaceBundleService({runHelper, invalidatePreview,
  stateDir = path.join(process.env.OPENCLAW_STATE_DIR || path.join(os.homedir(), '.openclaw'), '.ods-workspace-bundle-lanes')} = {}) {
  return async function execute(scope, supplied, signal) {
    const request = normalizeWorkspaceBundle(supplied);
    if (typeof scope?.workspaceRoot !== 'string' || !path.isAbsolute(scope.workspaceRoot)) throw fail('workspace-unavailable');
    const identity = hash(scope.workspaceRoot), queueKey = path.join(stateDir, identity);
    const previousJob = bundleQueues.get(queueKey) ?? Promise.resolve();
    const work = previousJob.catch(() => {}).then(async () => {
      if (signal?.aborted) throw fail('bundle-cancelled');
      privateDirectory(stateDir);
      const file = path.join(stateDir, identity + '.json');
      if (readPrivate(file)) throw fail('bundle-workspace-execution-unsettled');
      const generation = 'bundle-' + randomUUID().replaceAll('-','');
      const state = {schemaVersion:2, workspaceRoot:scope.workspaceRoot, generation,
        runId:scope.runId, sessionId:scope.sessionId, sessionKey:scope.sessionKey, status:'pending'};
      let fd;
      try { fd = fs.openSync(file, fs.constants.O_WRONLY | fs.constants.O_CREAT | fs.constants.O_EXCL | (fs.constants.O_NOFOLLOW ?? 0), 0o600); }
      catch (error) { if(error.code === 'EEXIST') throw fail('bundle-workspace-execution-unsettled'); throw error; }
      try { fs.writeFileSync(fd, JSON.stringify(state)); fs.fsyncSync(fd); } finally { fs.closeSync(fd); }
      let settled = false;
      try {
        if (signal?.aborted || !invalidatePreview(scope)) { settled = true; throw fail('bundle-run-unavailable'); }
        const result = await runHelper(scope, {request, generation}, signal, {
          onProcess: execution => savePrivate(file, {...state, execution}),
          onSettled: () => { settled = true; },
        });
        const receipt = helperReceipt(result, request, generation);
        return {...receipt, request,
          boundary:'Immutable derived bytes only. Use returned generation paths; fixed-path requests require ordinary file tools. No execution, publication, or task-correctness proof.'};
      } finally {
        // Failure to prove quiescence leaves a durable guard even after this
        // process exits. A retry must not overlap an uncertain older helper.
        if (settled && readPrivate(file)?.generation === generation) fs.unlinkSync(file);
      }
    });
    bundleQueues.set(queueKey, work);
    try { return await work; }
    finally { if (bundleQueues.get(queueKey) === work) bundleQueues.delete(queueKey); }
  };
}

export function createWorkspaceBundleTool(context, {admission, execute, scopeForContext}) {
  if (context?.agentId !== 'pixel') return null;
  return {name: WORKSPACE_BUNDLE_TOOL, label: 'Bundle existing workspace files',
    description: 'Create an immutable generation of exact UTF-8 source maps and raw copies without retyping contents. outputRoot selects a workspace directory; mappingPath and copyTo are filenames relative to the new generation. Source paths are workspace-relative. Every call returns a fresh generation path with the original keys and filenames for publication/download links; include an index.html copy when the generation should itself be browser-publishable. Never replaces prior files. If the owner requires a literal fixed path, use ordinary file tools for that request rather than silently substituting these version paths. Does not run tests or publish a preview.',
    parameters: {type: 'object', additionalProperties: false, required: ['files', 'mappingPath', 'outputRoot'], properties: {
      files: {type: 'array', minItems: 1, maxItems: 32, items: {type: 'object', additionalProperties: false,
        required: ['source', 'key', 'copyTo'], properties: {source: {type: 'string'}, key: {type: 'string'}, copyTo: {type: 'string'}}}},
      mappingPath: {type: 'string'}, outputRoot: {type:'string'},
    }},
    async execute(id, args, signal) {
      try {
        const request = normalizeWorkspaceBundle(args), active = admission.take(id, request, context);
        const scope = scopeForContext(active, context);
        const details = await execute(scope, request, signal);
        return {details, content: [{type: 'text', text: 'Immutable workspace bundle verified byte-for-byte. ' + details.boundary + '\n' + JSON.stringify(details)}]};
      } catch (error) {
        return {isError: true, details: {status: 'failed', readbackVerified: false,
          ...(Array.isArray(error?.written) ? {written: error.written} : {})}, content: [{type: 'text',
          text: 'Workspace bundle was not verified. ' + (/^[a-z-]{1,80}$/.test(error?.code ?? '') ? error.code : 'bundle-unavailable') +
            '. Preserve existing files; do not claim these artifacts passed verification.'}]};
      }
    }};
}
